v10: Update to ImageSharp v2

ronaldbarendse opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/11329. Also see the related discussion https://github.com/umbraco/Umbraco-CMS/discussions/11938 and the draft PR https://github.com/umbraco/Umbraco-CMS/pull/11991 trying to update v9 in a backwards compatible way (spoiler: we can't).

### Description
This PR updates the dependency versions of ImageSharp to 2.1.0 and ImageSharp.Web to 2.0.0-alpha.0.23 (currently waiting on the final release). Because these new major versions contain breaking changes, we can only update these for the upcoming Umbraco v10 release. Some of the breaking changes are:
- `CachedNameLength` has been renamed to `CacheHashLength`
- `IDictionary<string, string> commands` has been changed to `CommandCollection commands`
- A new `RequiresTrueColorPixelFormat()` method is added to `IImageWebProcessor`

ImageSharp/ImageSharp.Web also has some functional changes as well:
- The `PhysicalFileSystemProvider` doesn't use the `WebRootFileProvider` anymore (it now only supports actual physical files OOTB)
- The `ResizeWebProcessor` is now EXIF-orientation aware (as browsers now default to displaying images based on the EXIF orientation)
- The `PhysicalFileSystemCache` has a separate option for the cache folder depth (it previously used the file length to create nested cache folders)
- It adds support for WebP images and includes better memory management

--------------
To-do:
- Update to final ImageSharp.Web v2 release (and remove MyGet feed)
- Implement EXIF-orientation aware cropping in `CropWebProcessor` (waiting on PR https://github.com/SixLabors/ImageSharp.Web/pull/241 that adds an utility class to make this easier)
- Add WebP to default image file types/extensions?
  https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Core/Configuration/Models/ContentImagingSettings.cs#L26

JimBobSquarePants commented on src/Umbraco.Web.Common/ImageProviders/FileProviderImageProvider.cs:
```diff
@@ -0,0 +1,62 @@
+    /// <summary>
+    /// Returns images from an <see cref="IFileProvider"/> abstraction.
+    /// </summary>
+    public class FileProviderImageProvider : IImageProvider
```
I like the pattern you've used here. You might be able to convince me to add this upstream.

ronaldbarendse replied:
Would be great if you're willing to merge this into ImageSharp.Web, will create a PR for it 🎉 It can/should lose the virtual methods for better performance when not inherited and sealed, so you can also use this with a factory registration:
```c#
.AddProvider(provider => new FileProviderImageProvider(
    provider.GetRequiredService<IWebHostEnvironment>().WebRootFileProvider,
    provider.GetRequiredService<FormatUtilities>()))
```

JimBobSquarePants commented on src/Umbraco.Web.Common/ImageProviders/WebRootImageProvider.cs:
```diff
@@ -0,0 +1,20 @@
+        /// <param name="environment">The web host environment.</param>
+        /// <param name="formatUtilities">Contains various format helper methods based on the current configuration.</param>
+        public WebRootImageProvider(IWebHostEnvironment environment, FormatUtilities formatUtilities)
+            : base(environment?.WebRootFileProvider, formatUtilities)
```
Can the `IWebHostEnvironment ` be null? `WebRootFileProvider` definitely but I would have thought the entire application would fail without the environment no?

ronaldbarendse replied:
You're totally right, this can simply be updated to:
```suggestion
            : base(environment.WebRootFileProvider, formatUtilities)
```

JimBobSquarePants commented:
@Nuklon That was quick 😉 I published the blog post about 30 seconds ago!

Nuklon commented:
> @Nuklon That was quick 😉 I published the blog post about 30 seconds ago!

Saw it on NuGet :)

ronaldbarendse commented on src/Umbraco.Core/Configuration/Models/ContentImagingSettings.cs:
```diff
@@ -23,7 +23,7 @@ public class ContentImagingSettings
         };
 
-        internal const string StaticImageFileTypes = "jpeg,jpg,gif,bmp,png,tiff,tif";
+        internal const string StaticImageFileTypes = "jpeg,jpg,gif,bmp,png,tiff,tif,webp";
```
This setting doesn't seem to be used: the supported image file types are now retrieved from the `IImageUrlGenerator.SupportedImageFileTypes` property and the `ImageSharpImageUrlGenerator` implementation dynamically retrieves the supported ones from the ImageSharp configuration instead.

These image file types are then used to for more then just generating image URLs though (e.g. whether to extract image dimensions to auto-fill properties, disallow file uploads that are not images and automatically changing the media type from `File` to `Image`).

I suggest keeping this setting and create a separate PR to fix the mixed usage, so:
- `ContentImagingSettings.ImageFileTypes` is used to determine whether the uploaded file is a valid image and if it can automatically change the media type to `Image`;
- `IImageUrlGenerator.SupportedImageFileTypes` is used whether it can generate an image URL for a specified file;
- A new `IImageDimensionExtractor.SupportedImageFileTypes` property is added that exposes whether it can get dimensions for a specified file (and have the `ImageSharpDimensionExtractor` populate this the same way as the `ImageSharpImageUrlGenerator` currently does).

ronaldbarendse commented on src/Umbraco.Web.Common/Media/ImageSharpImageUrlGenerator.cs:
```diff
@@ -0,0 +1,100 @@
+
+            if (options.CacheBusterValue is string cacheBusterValue && !string.IsNullOrWhiteSpace(cacheBusterValue))
+            {
+                queryString.Add("rnd", cacheBusterValue);
```
I think `v` would be a much better (and shorter) query string key, as this isn't a random value we're adding, but a versioning value. Changing this doesn't impact the generation of images: it would only invalidate any images in your browser cache.

And that might actually be for the better, as you should also clear your existing ImageSharp cache as part of the migration, because newly generated images can be different because of the automatic EXIF rotation.

```suggestion
                queryString.Add("v", cacheBusterValue);
```

_This should then also be updated in the corresponding tests._

ronaldbarendse commented:
### Breaking changes
- Update dependency versions:
  - `SixLabors.ImageSharp` from 1.0.4 to 2.1.1 (see breaking changes in [v2.0.0 release notes](https://github.com/SixLabors/ImageSharp/releases/tag/v2.0.0));
  - `SixLabors.ImageSharp.Web` from 1.0.5 to 2.0.0 (see breaking changed in [release notes](https://github.com/SixLabors/ImageSharp.Web/releases/tag/v2.0.0));
- Renamed `CachedNameLength` property to `CacheHashLength` on `ImagingCacheSettings`;
- Moved `ImageSharpImageUrlGenerator` from project `Umbraco.Infrastructure` to `Umbraco.Web.Common` and updated the corresponding namespace and DI registration (from `AddCoreInitialServices()` to `AddUmbracoImageSharp()`);
- Moved ImageSharp configuration from the `AddUmbracoImageSharp()` extension method into separate `IConfigureOptions<>` implementations:
  - The middleware is configured in `ConfigureImageSharpMiddlewareOptions` (which also replaces `ImageSharpConfigurationOptions` that previously only set the default ImageSharp configuration);
  - The default physical cache is configured in `ConfigurePhysicalFileSystemCacheOptions`.

There are also some functional changes to be aware of:
- The default physical file cache has a different/optimized folder structure (less nesting and shorter file names): **make sure to delete your existing cache folder if you're upgrading** (located in `umbraco\Data\TEMP\MediaCache`);
- The cache key is now only generated based on the relative URL (it doesn't include the hostname anymore), so the same cached image is returned if you have multiple hostnames configured in Umbraco;
- Images are automatically rotated based on their EXIF orientation (another reason to remove your existing cached images);
- WebP support is added 🎉

bergmania commented on src/Umbraco.Infrastructure/DependencyInjection/UmbracoBuilder.CoreServices.cs:
```diff
@@ -200,7 +200,6 @@ public static IUmbracoBuilder AddCoreInitialServices(this IUmbracoBuilder builde
             // Add default ImageSharp configuration and service implementations
             builder.Services.AddSingleton(SixLabors.ImageSharp.Configuration.Default);
             builder.Services.AddSingleton<IImageDimensionExtractor, ImageSharpDimensionExtractor>();
```
Normally we add a NoopImpementation so you can boot Infrastructure without the web assemblies
